irohad/synchronizer/impl/synchronizer_impl.cpp

```diff
@@ -102,32 +104,33 @@ namespace iroha {
               my_height + 1,
               target_height,
               public_key);
-          auto network_chain =
-              block_loader_
-                  ->retrieveBlocks(my_height,
-                                   PublicKeyHexStringView{public_key})
-                  .tap([&my_height, &got_some_blocks_from_this_peer](
-                           const std::shared_ptr<shared_model::interface::Block>
-                               &block) {
-                    got_some_blocks_from_this_peer = true;
-                    my_height = block->height();
-                  });
-
-          if (validator_->validateAndApply(network_chain, *storage)) {
-            if (my_height >= target_height) {
-              // goto is alright to break out of nested loops:
-              // https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es76-avoid-goto
-              return mutable_factory_->commit(std::move(storage));
-            }
-            if (not got_some_blocks_from_this_peer) {
-              // if we got no new blocks from this peer we should switch to
-              // next peer
+          auto retrieve_blocks_result = block_loader_->retrieveBlocks(
+              my_height, PublicKeyHexStringView{public_key});
+
+          if (hasValue(retrieve_blocks_result)) {
+            auto network_chain_with_updates =
+                retrieve_blocks_result.assumeValue().tap(
```

Comment:
Need to catch on_error from observable, please check if tests are present for this case.

Reply:
Caught in mutable storage